### PR TITLE
Add `Bill` model implementation. 

### LINKED ISSUE #16: Implement Bill Model
### Common description 

Model of bill should contain:

* next fields to store data: 

	- how much every participant paid. The dictionary **[name: sum]** of `[String: Decimal]` type .

	- the overall sum of the bill. A `Decimal` value. 

	- the description of the bill. A `String` value. 

	- date. A value of `Date` type that reflects date in the user's current time zone.

	- id: `UUID` type. 

### Example of implementation

```swift
struct Bill {

	let participantSpent: [String: Decimal]
    
	let description: String
    
	let sumOfBill: Decimal
    
	let date: Date
    
	let id = UUID()
}
```

### Fields validations: 

- `participantSpent`:
	- `participantSpent.key`:
		- contain no greater than 50 characters
		- contain at least one character other than a space

	-  `participantSpent.value`:
		- can't be negative 
		- the maximum is 2 decimal places 

- `description`:
	- the maximum length is 50 characters.

* `sumOfBill`: 
	* the numbers must be greater than zero
	* `sumOfBill` is equal to the sum of the *participants's sums* 

- `date`: 
	- the date can't be in the future.

### PR DESCRIPTION
**Pull request by kirpichonok**
# What 

Added an implementation of the `Bill` model. 
 
# Why

It closes #16.

# How 

- Add type `Bill` with fields to save bill's data. 

- Added validation of the input data  to the `Bill` initialiser.

- The type `BillValidationError` adopted `Equatable` protocol to verify the received errors during testing.

- Added `BillTests` file with tests.

### CODE REVIEW

**Review by YuliaNastiushkina — COMMENTED** [review]
I have a few questions, please check.

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    /// A date for the bill that was set up by the user.
+    let date: Date
+
+    /// The unique identifier for the bill.
+    let id = UUID()
+
+    // TODO: Documentation for the rest of API
+    init(
+        participantSpents: [String: Decimal],
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:22]: You're right. Fixed

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        description: String,
+        sumOfBill: Decimal,
+        date: Date
+    ) throws {
+        try Bill.validateBill(participantSpent: participantSpents,
+                              description: description,
+                              sumOfBill: sumOfBill,
+                              date: date)
+
+        participantSpent = participantSpents
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:32]: Fixed

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        self.date = date
+    }
+
+    private static func validateBill(
+        participantSpent: [String: Decimal],
+        description _: String,
+        sumOfBill: Decimal,
+        date: Date
+    ) throws {
+        guard date < Date() else { throw BillValidationError.futureDate(date) }
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:44]: Fixed

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    private static func hasMoreThanTwoDecimalPlaces(_ number: Decimal) -> Bool {
+        let decimalDigits = number.description.components(separatedBy: ".")
+        return decimalDigits.count > 1 && decimalDigits[1].count > 2
+    }
+}
+
+extension Bill {
+    /// The allowed length of the description of the bill.
+    static let allowedDescriptionLength = 50
```
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:83]: Please review the rest of code.

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        guard name.count <= Bill.allowedParticipantNameLength else {
+            throw BillValidationError.tooLongParticipantName(name: name)
+        }
+
+        guard !name.trimmingCharacters(in: .whitespaces).isEmpty else {
+            throw BillValidationError.emptyParticipantName(name: name)
+        }
+    }
+
+    private static func hasMoreThanTwoDecimalPlaces(_ number: Decimal) -> Bool {
```
Comment (dersim-davaod) [inline JourneysSquad/JourneysSquad/Model/BillModel.swift:75]: The implementation is fragile. Can you improve it?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Please consider the following improvements:

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 6]
```diff
@@ -0,0 +1,87 @@
+import Foundation
+
+/// Bill with related data.
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:3]: Fixed

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+struct Bill {
+    /// How much each participant has spent within the bill.
+    let participantSpent: [String: Decimal]
+
+    /// The user's notes to the bill.
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:8]: Fixed

File: JourneysSquad/JourneysSquad/Model/BillModel.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:83]: Fixed

[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift:83]: Fixed

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,3 +9,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Build scheme for JourneySquad target.
 - `BillValidationError` type.
+- Add `Bill` model into the model layer - [#32](https://github.com/ios-course/link-team-project/pull/32)
```
Comment (dersim-davaod) [inline CHANGELOG.md:12]: Why you haven't added the `Add` prefix earlier?
Be consistent please: either use the prefix or not.

File: JourneysSquad/JourneysSquad/Model/BillValidationError.swift [hunk 9]
```diff
@@ -54,3 +54,40 @@ extension BillValidationError: LocalizedError {
         }
     }
 }
+
+extension BillValidationError: Equatable {
```
Comment (dersim-davaod) [inline JourneysSquad/JourneysSquad/Model/BillValidationError.swift:58]: Seems you can put the extension into the test target.

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                date: Date.now
+            )
+        ) { error in
+            XCTAssertEqual(error as? BillValidationError,
+                           BillValidationError.incorrectEstimatedSumOfTheBill)
+        }
+    }
+
+    func testInitThrowsTheDateIsInFutureErrorWhenTheInputDateIsInTheFuture() {
+        let date = Date.now + 10
```
Comment (dersim-davaod) [inline JourneysSquad/JourneysSquadTests/Model/BillTests.swift:230]: Magic numbers and non-clear semantics. Please improve using the approach we have discussed earlier.

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Please consider minor improvements for now:

File: CHANGELOG.md (hunk 8 above)
Comment (kirpichonok) [reply CHANGELOG.md:12]: Fixed

File: JourneysSquad/JourneysSquad/Model/BillValidationError.swift (hunk 9 above)
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillValidationError.swift:58]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 10 above)
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:230]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+                sumOfBill: sum,
+                date: Date.now
+            )
+        ) { error in
+            XCTAssertEqual(error as? BillValidationError,
+                           BillValidationError.invalidSumOfBill(sum: sum))
+        }
+    }
+
+    func testInitThrowsIncorrectEstimatedSumOfTheBilErrorlWhenSumOfBillIsNotEqualToSumOfPersonPaid() {
```
Comment (levvaa) [inline JourneysSquad/JourneysSquadTests/Model/BillTests.swift:215]: nit:

```suggestion
    func testInitThrowsIncorrectEstimatedSumOfTheBillErrorlWhenSumOfBillIsNotEqualToSumOfPersonPaid() {
```

**Review by levvaa — CHANGES_REQUESTED** [review]

File: JourneysSquad/JourneysSquad/Model/Billl.swift [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+    /// Create a new instance of bill.
+    ///
+    /// - Parameters:
+    ///    - personPaid: The list of the people and according amount of money that each of them paid.
+    ///    - description: The description of the bill.
+    ///    - sumOfBill: The sum of money that was spent by participants within the bill.
+    ///    - date: A date when the bill was created.
+    ///
+    /// - Throws: The `BillValidationError` if the bill can't be created or doesn't make sense.
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/Billl.swift:28]: I've update this part of documentation. Please, check out.

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+            case let .tooLongPersonName(name: name):
+                XCTAssertEqual(actualError,
+                               BillValidationError.tooLongPersonName(name: name))
+            default:
+                XCTFail("Unexpected error")
+            }
+        }
+    }
+
+    func testInitThrowsEmptyPesonNameErrorWhenOneOfNamesIsEmptyString() {
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:80]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+                                  date: Date.now)
+        let secondBill = try? Bill(personPaid: ["A": 1],
+                                   description: "Description",
+                                   sumOfBill: 1,
+                                   date: Date.now)
+
+        XCTAssertNotEqual(firstBill?.id, secondBill?.id)
+    }
+
+    func testInitThrowsTooLongPersonNameErrorWhenOneOfNamesIsLongerThenMaxPersonNameLength() {
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:39]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:80]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:39]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 11 above)
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:215]: Fixed

File: JourneysSquad/JourneysSquad/Model/Billl.swift
Comment (levvaa) [inline JourneysSquad/JourneysSquad/Model/Billl.swift:1]: Check the name of file

**Review by levvaa — CHANGES_REQUESTED** [review]

File: JourneysSquad/JourneysSquad/Model/Billl.swift
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/Billl.swift:1]: It's epic fail 😬
Fixed

**Review by levvaa — APPROVED** [review]
LGTM

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    func testInitThrowsTooLongPersonNameErrorWhenOneOfNamesIsLongerThanMaxPersonNameLength() {
+        let name = String(repeating: "A",
+                          count: Bill.maxPersonNameLength + 1)
+
+        XCTAssertThrowsError(
+            try Bill(
+                personPaid: [name: 1],
+                description: "Description",
```
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:46]: Please review the rest of code.

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Consider improvements to increase maintainability.

File: JourneysSquad/JourneysSquadTests/Extensions/BillValidationErrorExtension.swift [hunk 16]
```diff
@@ -0,0 +1,36 @@
+extension BillValidationError: Equatable {
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Extensions/BillValidationErrorExtension.swift:1]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+                date: Date.now
+            )
+        ) { error in
+            XCTAssertEqual(error as? BillValidationError,
+                           BillValidationError.emptyPersonName(name: name))
+        }
+    }
+
+    func testInitThrowsEmptyPesonNameErrorWhenOneOfNamesContainOnlySpaces() {
+        let name = " "
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:97]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+            )
+        ) { error in
+            XCTAssertEqual(error as? BillValidationError,
+                           BillValidationError.emptyPersonName(name: name))
+        }
+    }
+
+    func testInitThrowsEmptyPesonNameErrorWhenOneOfNamesContainsOnlySpacesAndAnotherNameIsEmpty() {
+        let nameOne = ""
+        let nameTwo = " "
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:114]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:46]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    func testInitThrowsTooLongPersonNameErrorWhenOneOfNamesIsLongerThanMaxPersonNameLength() {
+        let name = String(repeating: "A",
+                          count: Bill.maxPersonNameLength + 1)
+
+        XCTAssertThrowsError(
+            try Bill(
+                personPaid: [name: 1],
+                description: "Description",
+                sumOfBill: 1,
+                date: Date.now
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:48]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+            let actualError = error as? BillValidationError
+
+            switch actualError {
+            case let .tooLongPersonName(name: name):
+                XCTAssertEqual(actualError,
+                               BillValidationError.tooLongPersonName(name: name))
+            default:
+                XCTFail("Unexpected error")
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:77]: I've done another way. Look, please.

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+        let secondBill = try? Bill(personPaid: ["A": 1],
+                                   description: "Description",
+                                   sumOfBill: 1,
+                                   date: Date.now)
+
+        XCTAssertNotEqual(firstBill?.id, secondBill?.id)
+    }
+
+    func testInitThrowsTooLongPersonNameErrorWhenOneOfNamesIsLongerThanMaxPersonNameLength() {
+        let name = String(repeating: "A",
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift:40]: Fixed

File: JourneysSquad/JourneysSquadTests/Extensions/BillValidationErrorExtension.swift [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    static func == (_ lhs: BillValidationError, _ rhs: BillValidationError) -> Bool {
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Extensions/BillValidationErrorExtension.swift:2]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+            try Bill(personPaid: ["A": 1],
+                     description: "Description",
+                     sumOfBill: 1,
+                     date: Date.now)
+        )
+    }
+
+    func testInitCreatesAnObjectWhosePropertiesHaveTheValuesOfThePassedParameters() {
+        let date = Date.now
+        let bill = try? Bill(personPaid: ["A": 1],
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift]: Fixed

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Please address all the comments above before re-requesting review.

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+                date: Date.now
+            )
+        ) { error in
+            let actualError = error as? BillValidationError
+
+            switch actualError {
+            case let .emptyPersonName(name: name):
+                XCTAssertEqual(actualError, BillValidationError.emptyPersonName(name: name))
+            default:
+                XCTFail("Unexpected error")
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+            }
+        }
+    }
+
+    func testInitThrowsNegativeAmountOfMoneyErrorWhenThePersonPaidAmountIsLessThenZero() {
+        let amount: Decimal = -1
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+                sumOfBill: 1,
+                date: Date.now
+            )
+        ) { error in
+            XCTAssertEqual(error as? BillValidationError,
+                           BillValidationError.emptyPersonName(name: name))
+        }
+    }
+
+    func testInitThrowsEmptyPesonNameErrorWhenOneOfNamesContainsOnlySpacesAndAnotherNameIsEmpty() {
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift]: I have rewritten the test. I think the name now corresponds to content.

File: JourneysSquad/JourneysSquad/Model/Bill.swift [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+        self.date = date
+    }
+
+    private static func validateBill(
+        personPaid: [String: Decimal],
+        description _: String,
+        sumOfBill: Decimal,
+        date: Date
+    ) throws {
+        guard date <= Date() else { throw BillValidationError.theDateIsInFuture(date) }
```
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/Bill.swift:53]: I think that `guard` operator is more suitable for early exit or error throwing.

**Comment by kirpichonok** [comment]
Sorry, I've missed hidden conversations.

File: JourneysSquad/JourneysSquad/Model/Bill.swift [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+    /// The overall sum of money that was spent within the bill.
+    let sumOfBill: Decimal
+
+    /// A date when the bill was created.
+    let date: Date
+
+    /// The unique identifier for the bill.
+    let id = UUID()
+
+    /// Create a new instance of bill.
```
Comment (levvaa) [inline JourneysSquad/JourneysSquad/Model/Bill.swift:20]: nit:

```suggestion
    /// Creates a new instance of bill.
```

**Review by levvaa — APPROVED** [review]
LGTM
Just pay attention to nit

File: JourneysSquad/JourneysSquad/Model/Bill.swift (hunk 28 above)
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/Bill.swift:20]: Fixed

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift]: Still see the same issue.
Let me know when the new version arrives.

File: JourneysSquad/JourneysSquad/Model/BillModel.swift [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+
+        participantSpent = participantSpents
+        self.description = description
+        self.sumOfBill = sumOfBill
+        self.date = date
+    }
+
+    private static func validateBill(
+        participantSpent: [String: Decimal],
+        description _: String,
```
[the remark this answers is not in the document]
Comment (YuliaNastiushkina) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift]: Please, review the rest of code.
Line 49, for example.

**Review by YuliaNastiushkina — COMMENTED** [review]
Great job, but this PR so huge.
It difficult to review, so I cannot guarantee you that the review was of the best quality.
Please try not to make the PR so big next time.

I have some questions, please check.

File: JourneysSquad/JourneysSquadTests/Model/BillTests.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquadTests/Model/BillTests.swift]:
```swift
        do {
            _ = try Bill(
                personPaid: [nameThatExceedsMaxLengthOne: 1, nameThatExceedsMaxLengthTwo: 1],
                description: "",
                sumOfBill: 1,
                date: .now
            )
        } catch BillValidationError.tooLongPersonName {
            XCTAssert(true)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
```
I founded one more variant. 

@dersim-davaod

**Review by dersim-davaod — APPROVED** [review]
LGTM

**Review by YuliaNastiushkina — COMMENTED** [review]
Please address all the comments above before re-requesting review.

File: JourneysSquad/JourneysSquad/Model/BillModel.swift (hunk 29 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply JourneysSquad/JourneysSquad/Model/BillModel.swift]: Fixed

**Review by YuliaNastiushkina — APPROVED** [review]
LGTM

**Review by levvaa — APPROVED** [review]
LGTM

**Pull request merged**